File: TripKitAndroid/src/main/java/com/skedgo/tripkit/bookingproviders/BookingResolverImpl.java
package com.skedgo.tripkit.bookingproviders;

import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.res.Resources;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.skedgo.tripkit.BookingAction;
import com.skedgo.tripkit.ExternalActionParams;
import com.skedgo.tripkit.BookingAction;
import com.skedgo.tripkit.ExternalActionParams;
import io.reactivex.Observable;
import io.reactivex.functions.Function;
import skedgo.tripkit.geocoding.ReverseGeocodable;

import java.util.HashMap;
import java.util.Map;

public final class BookingResolverImpl implements BookingResolver {
  private Map<String, BookingResolver> resolverMap;

  public BookingResolverImpl(
      @NonNull Resources resources,
      @NonNull final PackageManager packageManager,
      @NonNull ReverseGeocodable geocoderFactory) {
    final Function<String, Boolean> isPackageInstalled = new Function<String, Boolean>() {
      @Override public Boolean apply(String packageName) {
        try {
          packageManager.getPackageInfo(packageName, PackageManager.GET_ACTIVITIES);
          return true;
        } catch (PackageManager.NameNotFoundException ignored) {
        }
        return false;
      }
    };

    final Function<String, Intent> getAppIntent = new Function<String, Intent>() {
      @Override public Intent apply(String packageName) {
        return packageManager.getLaunchIntentForPackage(packageName);
      }
    };

    resolverMap = new HashMap<>(8);
    resolverMap.put("gocatch", new GoCatchBookingResolver(resources, isPackageInstalled, geocoderFactory));
    resolverMap.put("ingogo", new IngogoBookingResolver(resources, isPackageInstalled));
    resolverMap.put("mtaxi", new MTaxiBookingResolver(isPackageInstalled, getAppIntent));
    resolverMap.put("uber", new UberBookingResolver(isPackageInstalled, getAppIntent));
    resolverMap.put("lyft", new LyftBookingResolver(resources, isPackageInstalled));
    resolverMap.put("flitways", new FlitWaysBookingResolver(geocoderFactory));
    resolverMap.put("tel:", new TelBookingResolver(resources));
    resolverMap.put("sms:", new SmsBookingResolver());
    resolverMap.put("http", new WebBookingResolver(resources));
  }

  @Override
  public Observable<BookingAction> performExternalActionAsync(ExternalActionParams params) {
    final String externalAction = params.action();
    final BookingResolver resolver = getBookingResolver(externalAction);
    if (resolver != null) {
      return resolver.performExternalActionAsync(params);
    } else {
      return Observable.error(new UnsupportedOperationException("Strange action: " + externalAction));
    }
  }

  @Nullable @Override public String getTitleForExternalAction(String externalAction) {
    BookingResolver resolver = getBookingResolver(externalAction);
    if (resolver != null) {
      return resolver.getTitleForExternalAction(externalAction);
    } else {
      return null;
    }
  }

  private BookingResolver getBookingResolver(String externalAction) {
    if (externalAction.startsWith("lyft")) {
      return resolverMap.get("lyft");
    } else if (externalAction.startsWith("http")) {
      return resolverMap.get("http");
    } else if (externalAction.startsWith("tel:")) {
      return resolverMap.get("tel:");
    } else if (externalAction.startsWith("sms:")) {
      return resolverMap.get("sms:");
    } else if (resolverMap.containsKey(externalAction)) {
      return resolverMap.get(externalAction);
    } else {
      return null;
    }
  }
}

File: TripKitAndroid/src/main/java/com/skedgo/tripkit/bookingproviders/BookingResolverImpl.kt
package com.skedgo.tripkit.bookingproviders

import android.content.Intent
import android.content.pm.PackageManager
import android.content.pm.PackageManager.NameNotFoundException
import android.content.res.Resources
import com.skedgo.tripkit.BookingAction
import com.skedgo.tripkit.ExternalActionParams
import com.skedgo.tripkit.geocoding.ReverseGeocodable
import io.reactivex.Observable
import io.reactivex.functions.Function

class BookingResolverImpl(
    resources: Resources,
    packageManager: PackageManager,
    geocoderFactory: ReverseGeocodable
) : BookingResolver {
    private val resolverMap: MutableMap<String?, BookingResolver>

    init {
        val isPackageInstalled: Function<String, Boolean> = object : Function<String, Boolean> {
            override fun apply(packageName: String): Boolean {
                try {
                    packageManager.getPackageInfo(packageName, PackageManager.GET_ACTIVITIES)
                    return true
                } catch (ignored: NameNotFoundException) {
                }
                return false
            }
        }

        val getAppIntent: Function<String, Intent> =
            Function<String, Intent> { packageName -> packageManager.getLaunchIntentForPackage(packageName) }

        resolverMap = HashMap(8)
        resolverMap["gocatch"] =
            GoCatchBookingResolver(resources, isPackageInstalled, geocoderFactory)
        resolverMap["ingogo"] = IngogoBookingResolver(resources, isPackageInstalled)
        resolverMap["mtaxi"] = MTaxiBookingResolver(isPackageInstalled, getAppIntent)
        resolverMap["uber"] = UberBookingResolver(isPackageInstalled, getAppIntent)
        resolverMap["lyft"] = LyftBookingResolver(resources, isPackageInstalled)
        resolverMap["flitways"] = FlitWaysBookingResolver(geocoderFactory)
        resolverMap["tel:"] = TelBookingResolver(resources)
        resolverMap["sms:"] = SmsBookingResolver()
        resolverMap["http"] = WebBookingResolver(resources)
    }

    override fun performExternalActionAsync(params: ExternalActionParams): Observable<BookingAction> {
        val externalAction = params.action()
        val resolver = getBookingResolver(externalAction)
        return resolver?.performExternalActionAsync(params)
            ?: Observable.error(
                UnsupportedOperationException(
                    "Strange action: $externalAction"
                )
            )
    }

    override fun getTitleForExternalAction(externalAction: String): String? {
        val resolver = getBookingResolver(externalAction)
        return resolver?.getTitleForExternalAction(externalAction)
    }

    private fun getBookingResolver(externalAction: String?): BookingResolver? {
        return if (externalAction!!.startsWith("lyft")) {
            resolverMap["lyft"]
        } else if (externalAction.startsWith("http")) {
            resolverMap["http"]
        } else if (externalAction.startsWith("tel:")) {
            resolverMap["tel:"]
        } else if (externalAction.startsWith("sms:")) {
            resolverMap["sms:"]
        } else if (resolverMap.containsKey(externalAction)) {
            resolverMap[externalAction]
        } else {
            null
        }
    }
}